File: frontend/src/lib/api/services/violationService.ts
/**
 * Violation Service
 * 
 * Business logic for violation management
 * Handles violation-related API calls, statistics, and data transformation
 * 
 * @module lib/api/services/violationService
 */

import { httpClient } from '../httpClient'
import type { PaginatedResponse } from '../types'

/**
 * Violation Entity
 */
export interface Violation {
  id: number
  camera_id: number
  domain_id: number
  timestamp: string
  person_bbox: { x: number; y: number; w: number; h: number }
  detected_ppe: Array<{ type: string; confidence: number }>
  missing_ppe: Array<{ type: string; required: boolean; priority: number }>
  track_id?: number
  confidence: number
  severity: 'critical' | 'high' | 'medium' | 'low'
  status: 'open' | 'in_progress' | 'closed' | 'false_positive'
  assigned_to?: string
  notes?: string
  corrective_action?: string
  snapshot_path?: string
  frame_snapshot?: string
  acknowledged: boolean
  acknowledged_by?: string
  acknowledged_at?: string
  created_at: string
}

/**
 * Violation Statistics
 */
export interface ViolationStatistics {
  total: number
  critical: number
  high: number
  medium: number
  low: number
  by_ppe_type: Record<string, number>
  compliance_rate: number
}

/**
 * Violation Filter Parameters
 */
export interface ViolationFilters {
  domain_id?: number
  camera_id?: number
  status?: 'open' | 'in_progress' | 'closed' | 'false_positive'
  severity?: 'critical' | 'high' | 'medium' | 'low'
  missing_ppe_type?: string
  start_date?: string
  end_date?: string
  skip?: number
  limit?: number
  acknowledged?: boolean // Legacy field
}

/**
 * Violation Create Payload
 */
export interface ViolationCreatePayload {
  camera_id: number
  domain_id: number
  timestamp?: string
  person_bbox: { x: number; y: number; w: number; h: number }
  detected_ppe: Array<{ type: string; confidence: number }>
  missing_ppe: Array<{ type: string; required: boolean; priority: number }>
  confidence: number
  severity?: 'critical' | 'high' | 'medium' | 'low'
  frame_snapshot?: string
}

/**
 * Violation Update Payload
 */
export interface ViolationUpdatePayload {
  status?: 'open' | 'in_progress' | 'closed' | 'false_positive'
  assigned_to?: string
  notes?: string
  corrective_action?: string
  acknowledged?: boolean // Legacy field
  acknowledged_by?: string // Legacy field
}

/**
 * Violation Service
 * 
 * Encapsulates all violation-related API operations
 */
export class ViolationService {
  private readonly basePath = '/violations'

  /**
   * Get violations with filtering and pagination
   */
  async getAll(filters: ViolationFilters = {}): Promise<PaginatedResponse<Violation>> {
    const params: Record<string, any> = {}
    
    if (filters.domain_id) params.domain_id = filters.domain_id
    if (filters.camera_id) params.camera_id = filters.camera_id
    if (filters.status) params.status = filters.status
    if (filters.severity) params.severity = filters.severity
    if (filters.missing_ppe_type) params.missing_ppe_type = filters.missing_ppe_type
    if (filters.start_date) params.start_date = filters.start_date
    if (filters.end_date) params.end_date = filters.end_date
    if (filters.acknowledged !== undefined) params.acknowledged = filters.acknowledged
    params.skip = filters.skip || 0
    params.limit = filters.limit || 50

    return httpClient.get<PaginatedResponse<Violation>>(this.basePath, { params })
  }

  /**
   * Get violation by ID
   */
  async getById(violationId: number): Promise<Violation> {
    return httpClient.get<Violation>(`${this.basePath}/${violationId}`)
  }

  /**
   * Get violation statistics
   */
  async getStatistics(
    domainId?: number,
    startDate?: string,
    endDate?: string
  ): Promise<ViolationStatistics> {
    const params: Record<string, any> = {}
    if (domainId) params.domain_id = domainId
    if (startDate) params.start_date = startDate
    if (endDate) params.end_date = endDate

    return httpClient.get<ViolationStatistics>(`${this.basePath}/stats`, { params })
  }

  /**
   * Create a new violation
   * 
   * Typically called by ML Engine when a violation is detected
   */
  async create(payload: ViolationCreatePayload): Promise<Violation> {
    // Set default timestamp if not provided
    if (!payload.timestamp) {
      payload.timestamp = new Date().toISOString()
    }

    // Calculate severity if not provided
    if (!payload.severity) {
      payload.severity = this.calculateSeverity(payload.missing_ppe)
    }

    return httpClient.post<Violation>(this.basePath, payload)
  }

  /**
   * Update a violation
   * 
   * Typically used for acknowledgment or adding notes
   */
  async update(violationId: number, payload: ViolationUpdatePayload): Promise<Violation> {
    return httpClient.put<Violation>(`${this.basePath}/${violationId}`, payload)
  }

  /**
   * Acknowledge a violation
   */
  async acknowledge(violationId: number, acknowledgedBy?: string): Promise<Violation> {
    return this.update(violationId, {
      acknowledged: true,
      acknowledged_by: acknowledgedBy,
    })
  }

  /**
   * Calculate violation severity based on missing PPE
   * 
   * Business logic: Critical if required PPE missing, High if recommended missing
   */
  private calculateSeverity(missingPPE: Array<{ required: boolean; priority: number }>): 'critical' | 'high' | 'medium' | 'low' {
    if (missingPPE.length === 0) return 'low'

    const hasRequired = missingPPE.some((ppe) => ppe.required)
    const hasHighPriority = missingPPE.some((ppe) => ppe.priority === 1)

    if (hasRequired || hasHighPriority) return 'critical'
    if (missingPPE.length >= 2) return 'high'
    return 'medium'
  }
}

// Export singleton instance
export const violationService = new ViolationService()

File: backend/services/violation_service.py
"""
Violation Service
Business logic layer for violation management
Follows SOLID principles and clean architecture
"""

from typing import Optional, List, Tuple
from datetime import datetime
from sqlalchemy.ext.asyncio import AsyncSession

from backend.database import crud, schemas
from backend.database.models import Violation, ViolationSeverity
from backend.utils.logger import logger
from backend.utils.snapshots import save_snapshot_image


class ViolationService:
    """
    Service class for violation business logic
    
    Responsibilities:
    - Validate violation data
    - Calculate severity
    - Handle violation creation/update logic
    - Coordinate between API and database layers
    """
    
    def __init__(self, db: AsyncSession):
        """
        Initialize violation service
        
        Args:
            db: Database session (dependency injection)
        """
        self.db = db
    
    async def get_violations(
        self,
        filters: schemas.ViolationFilterParams
    ) -> Tuple[List[Violation], int]:
        """
        Get violations with filtering and pagination
        
        Args:
            filters: Filter parameters
            
        Returns:
            Tuple of (violations list, total count)
        """
        logger.debug(f"Fetching violations with filters: {filters}")
        violations, total = await crud.get_violations(self.db, filters)
        logger.info(f"Retrieved {len(violations)} violations (total: {total})")
        return violations, total
    
    async def get_violation_by_id(self, violation_id: int) -> Optional[Violation]:
        """
        Get violation by ID
        
        Args:
            violation_id: Violation ID
            
        Returns:
            Violation if found, None otherwise
        """
        logger.debug(f"Fetching violation {violation_id}")
        violation = await crud.get_violation_by_id(self.db, violation_id)
        if violation:
            logger.debug(f"Violation {violation_id} found")
        else:
            logger.warning(f"Violation {violation_id} not found")
        return violation
    
    async def create_violation(
        self,
        violation_data: schemas.ViolationCreate
    ) -> Violation:
        """
        Create a new violation
        
        Business logic:
        - Validate camera and domain exist
        - Calculate severity if not provided
        - Create violation record
        
        Args:
            violation_data: Violation creation data
            
        Returns:
            Created violation
            
        Raises:
            ValueError: If camera or domain not found
        """
        logger.info(f"Creating violation for camera {violation_data.camera_id}")
        
        # Validate camera exists
        camera = await crud.get_camera_by_id(self.db, violation_data.camera_id)
        if not camera:
            error_msg = f"Camera {violation_data.camera_id} not found"
            logger.error(error_msg)
            raise ValueError(error_msg)
        
        # Validate domain exists
        domain = await crud.get_domain_by_id(self.db, violation_data.domain_id)
        if not domain:
            error_msg = f"Domain {violation_data.domain_id} not found"
            logger.error(error_msg)
            raise ValueError(error_msg)
        
        # Calculate severity if not provided
        if not violation_data.severity:
            violation_data.severity = self._calculate_severity(
                violation_data.missing_ppe
            )
            logger.debug(f"Calculated severity: {violation_data.severity}")
        
        # Persist snapshot if provided
        if violation_data.frame_snapshot:
            snapshot_path = save_snapshot_image(violation_data.frame_snapshot)
            violation_data.snapshot_path = snapshot_path
            violation_data.frame_snapshot = None  # do not store base64 in DB

        # Create violation
        violation = await crud.create_violation(self.db, violation_data)
        logger.info(f"Violation {violation.id} created successfully")
        
        return violation
    
    async def update_violation(
        self,
        violation_id: int,
        update_data: schemas.ViolationUpdate
    ) -> Optional[Violation]:
        """
        Update a violation
        
        Args:
            violation_id: Violation ID
            update_data: Update data
            
        Returns:
            Updated violation if found, None otherwise
        """
        logger.info(f"Updating violation {violation_id}")
        
        violation = await crud.update_violation(
            self.db,
            violation_id,
            update_data
        )
        
        if violation:
            logger.info(f"Violation {violation_id} updated successfully")
        else:
            logger.warning(f"Violation {violation_id} not found for update")
        
        return violation
    
    async def acknowledge_violation(
        self,
        violation_id: int,
        acknowledged_by: str,
        notes: Optional[str] = None
    ) -> Optional[Violation]:
        """
        Acknowledge a violation
        
        Args:
            violation_id: Violation ID
            acknowledged_by: Username of person acknowledging
            notes: Optional notes
            
        Returns:
            Updated violation if found, None otherwise
        """
        logger.info(f"Acknowledging violation {violation_id} by {acknowledged_by}")
        
        update_data = schemas.ViolationUpdate(
            acknowledged=True,
            acknowledged_by=acknowledged_by,
            notes=notes
        )
        
        return await self.update_violation(violation_id, update_data)
    
    async def get_statistics(
        self,
        domain_id: Optional[int] = None,
        start_date: Optional[datetime] = None,
        end_date: Optional[datetime] = None
    ) -> dict:
        """
        Get violation statistics
        
        Args:
            domain_id: Optional domain filter
            start_date: Optional start date filter
            end_date: Optional end date filter
            
        Returns:
            Statistics dictionary
        """
        logger.debug(f"Fetching statistics for domain {domain_id}")
        stats = await crud.get_violation_stats(
            self.db,
            domain_id,
            start_date,
            end_date
        )
        logger.debug(f"Statistics retrieved: {stats}")
        return stats
    
    def _calculate_severity(
        self,
        missing_ppe: List[dict]
    ) -> ViolationSeverity:
        """
        Calculate violation severity based on missing PPE
        
        Business logic:
        - Critical: Required PPE missing or high priority missing
        - High: Multiple recommended PPE missing
        - Medium: Single recommended PPE missing
        - Low: No missing PPE (shouldn't happen, but handle gracefully)
        
        Args:
            missing_ppe: List of missing PPE items with required/priority flags
            
        Returns:
            Calculated severity level
        """
        if not missing_ppe:
            return ViolationSeverity.LOW
        
        # Check for required PPE
        has_required = any(
            item.get('required', False) for item in missing_ppe
        )
        
        # Check for high priority
        has_high_priority = any(
            item.get('priority', 3) == 1 for item in missing_ppe
        )
        
        if has_required or has_high_priority:
            return ViolationSeverity.CRITICAL
        
        # Multiple missing items
        if len(missing_ppe) >= 2:
            return ViolationSeverity.HIGH
        
        # Single missing item
        return ViolationSeverity.MEDIUM

